pgmax/groups/variables.py

```diff
@@ -22,27 +24,62 @@ class NDVariableArray(groups.VariableGroup):
             the notion of a NumPy ndarray shape)
     """
 
-    num_states: int
     shape: Tuple[int, ...]
+    num_states: np.ndarray
+
+    def __post_init__(self):
+        if np.isscalar(self.num_states) and np.issubdtype(type(np.int64(10)), int):
+            num_states = np.full(self.shape, fill_value=self.num_states)
+            object.__setattr__(self, "num_states", num_states)
+        elif isinstance(self.num_states, np.ndarray) and np.issubdtype(
+            self.num_states.dtype, int
+        ):
+            if self.num_states.shape != self.shape:
+                raise ValueError("Should be same shape")
+        else:
+            raise ValueError("num_states entries should be of type np.int")
+
+        random_hash = random.randint(0, 2**63)
+        object.__setattr__(self, "random_hash", random_hash)
+
+    def __hash__(self):
+        return self.random_hash
+
+    def __eq__(self, other):
+        return hash(self) == hash(other)
 
-    def _get_names_to_variables(
-        self,
-    ) -> OrderedDict[Union[int, Tuple[int, ...]], nodes.Variable]:
-        """Function that generates a dictionary mapping names to variables.
+    def __lt__(self, other):
+        return hash(self) < hash(other)
+
+    def __getitem__(self, val):
+        # Relies on numpy indexation to throw IndexError if val is out-of-bounds
+        result = (self.variable_names[val], self.num_states[val])
+        if isinstance(val, slice):
+            return tuple(zip(result))
+        return result
+
+    @cached_property
+    def variables(self) -> List[Tuple]:
+        """Function that returns the list of all variables in the VariableGroup.
+        Each variable is represented by a tuple of the form (variable hash, number of states)
 
         Returns:
-            a dictionary mapping all possible names to different variables.
+            List of variables in the VariableGroup
         """
-        names_to_variables: OrderedDict[
-            Union[int, Tuple[int, ...]], nodes.Variable
-        ] = collections.OrderedDict()
-        for name in itertools.product(*[list(range(k)) for k in self.shape]):
-            if len(name) == 1:
-                names_to_variables[name[0]] = nodes.Variable(self.num_states)
-            else:
-                names_to_variables[name] = nodes.Variable(self.num_states)
+        vars_names = self.variable_names.flatten()
+        vars_num_states = self.num_states.flatten()
+        return list(zip(vars_names, vars_num_states))
```

Comment:
This might be slow (since it looks like a loop)? Do we really need all of `variables`, `variable_names` and `num_states` separately?

Reply:
For example is there anything wrong to just represent `variables` as a numpy array (so that `variable_names` is just `variables[:, 0]` and `num_states` is just `variables[:, 1]`? Is the issue that we want to keep individual variables as tuples?

Reply:
I have played with this and it turns out that `list(zip(vars_names, vars_num_states))` is the fastest way to keep individual variables as tuples